### PR TITLE
introduce wait_timeout for topics, services, actions, tfs and diagnostics

### LINKED ISSUE #184: timeout concept for wait_for features
make recording fail with exception/timeout message when timeout is exceeded...

needed for:
- wait_for_action
- wait_for_topic
- wait_for_service
- wait_for_diagnostics

### PR DESCRIPTION
**Pull request by floweisshardt**
fixes #184

### CODE REVIEW

File: atf_core/src/atf_core/recorder.py [hunk 1]
```diff
@@ -127,6 +150,12 @@ def __init__(self, test):
 
         rospy.loginfo("ATF recorder: started!")
 
+    def check_for_timeout(self, start_time, timeout):
```
Comment (fmessmer) [inline atf_core/src/atf_core/recorder.py]: this function should accept an optional `msg` field - which can be used for additional info (such as last diagnostics)

File: atf_core/src/atf_core/recorder.py [hunk 2]
```diff
@@ -92,21 +103,33 @@ def __init__(self, test):
...
+                        pass
+                    rospy.logdebug("... waiting since %.1f sec for transformation from '%s' to '%s' to become available ..."%((rospy.Time.now() - start_time).to_sec(), root_frame, measured_frame))
                 rospy.loginfo("... transformation from '%s' to '%s' available.", root_frame, measured_frame)
 
         if test.robot_config != None and 'wait_for_diagnostics' in test.robot_config and test.robot_config["wait_for_diagnostics"]:
             rospy.loginfo("Waiting for diagnostics to become OK ...")
             r = rospy.Rate(100)
             while not rospy.is_shutdown() and self.diagnostics != None and self.diagnostics.level != 0:
-                rospy.logdebug("... waiting for diagnostics to become OK ...")
+                self.check_for_timeout(start_time, wait_timeout)
```
Comment (fmessmer) [inline atf_core/src/atf_core/recorder.py]:
```suggestion
                self.check_for_timeout(start_time, wait_timeout, "latest diagnostics: {}".format(self.diagnostics))
```

**Review by fmessmer — CHANGES_REQUESTED** [review]

**Comment by fmessmer** [comment]
`wait_for_topic` actually throws the ROS exception - I guess it's the same for `wait_for_service`
```
timeout exceeded while waiting for message on topic /i_do_not_exist
```
rather than the `ATFRecorderError`

File: atf_core/src/atf_core/recorder.py [hunk 3]
```diff
@@ -60,24 +61,34 @@ def __init__(self, test):
...
+            if test.robot_config["wait_timeout"] > 0:
+                wait_timeout = test.robot_config["wait_timeout"]
+                rospy.loginfo("wait_timeout is set to %s", str(wait_timeout))
+
         if test.robot_config != None and 'wait_for_topics' in test.robot_config:
             for topic in test.robot_config["wait_for_topics"]:
+                self.check_for_timeout(start_time, wait_timeout)
                 rospy.loginfo("Waiting for topic '%s'...", topic)
-                rospy.wait_for_message(topic, rospy.AnyMsg)
+                rospy.wait_for_message(topic, rospy.AnyMsg, wait_timeout)
```
Comment (fmessmer) [inline atf_core/src/atf_core/recorder.py]: when adding the timeout here, the above call to `check_for_timeout` is obsolete

File: atf_core/src/atf_core/recorder.py [hunk 4]
```diff
@@ -60,24 +61,34 @@ def __init__(self, test):
...
                 rospy.loginfo("... got message on topic '%s'.", topic)
 
         if test.robot_config != None and 'wait_for_services' in test.robot_config:
             for service in test.robot_config["wait_for_services"]:
+                self.check_for_timeout(start_time, wait_timeout)
                 rospy.loginfo("Waiting for service '%s'...", service)
-                rospy.wait_for_service(service)
+                rospy.wait_for_service(service, wait_timeout)
```
Comment (fmessmer) [inline atf_core/src/atf_core/recorder.py]: when adding the timeout here, the above call to `check_for_timeout` is obsolete

File: atf_core/src/atf_core/recorder.py [hunk 5]
```diff
@@ -60,24 +61,34 @@ def __init__(self, test):
...
                 rospy.loginfo("... service '%s' available.", service)
 
         if test.robot_config != None and 'wait_for_actions' in test.robot_config:
             for action in test.robot_config["wait_for_actions"]:
+                self.check_for_timeout(start_time, wait_timeout)
                 rospy.loginfo("Waiting for action '%s'...", action)
 
                 # wait for action status topic
-                rospy.wait_for_message(action + "/status", rospy.AnyMsg)
+                rospy.wait_for_message(action + "/status", rospy.AnyMsg, wait_timeout)
```
Comment (fmessmer) [inline atf_core/src/atf_core/recorder.py]: when adding the timeout here, the above call to `check_for_timeout` is obsolete

File: atf_core/src/atf_core/recorder.py [hunk 6]
```diff
@@ -92,21 +103,33 @@ def __init__(self, test):
                 client = SimpleActionClient(action, roslib.message.get_message_class(topic_type))
 
                 # wait for action server
-                client.wait_for_server()
+                if wait_timeout == None:
+                    client.wait_for_server()
+                else:
+                    client.wait_for_server(rospy.Duration(wait_timeout))
```
Comment (fmessmer) [inline atf_core/src/atf_core/recorder.py]: why is this if-else case only done here?

**Review by fmessmer — CHANGES_REQUESTED** [review]

**Comment by fmessmer** [comment]
I'm testing contributions wrt my comments in https://github.com/floweisshardt/atf/pull/186

**Comment by fmessmer** [comment]
I see a lot of these errors in the `atf` tests - seems something is wrong with the `create_subscriber` function of the publishers:
```
 test 24
        Start 24: _ctest_atf_test_rostest_test_generated_atf_test_generation_config_yaml_recording_ts1_c0_r0_e0_s0_1.test
  
  24: Test command: /root/target_ws/build/atf_test/catkin_generated/env_cached.sh "/usr/bin/python3" "/opt/ros/noetic/share/catkin/cmake/test/run_tests.py" "/root/target_ws/build/atf_test/test_results/atf_test/rostest-test_generated_atf_test_generation_config_yaml_recording_ts1_c0_r0_e0_s0_1.xml" "--return-code" "/usr/bin/python3 /opt/ros/noetic/share/rostest/cmake/../../../bin/rostest --pkgdir=/root/target_ws/src/atf/atf_test --package=atf_test --results-filename test_generated_atf_test_generation_config_yaml_recording_ts1_c0_r0_e0_s0_1.xml --results-base-dir "/root/target_ws/build/atf_test/test_results" /root/target_ws/build/atf_test/test_generated/atf_test_generation_config_yaml/recording_ts1_c0_r0_e0_s0_1.test "
  24: Test timeout computed to be: 10000000
  24: -- run_tests.py: execute commands
  24:   /usr/bin/python3 /opt/ros/noetic/share/rostest/cmake/../../../bin/rostest --pkgdir=/root/target_ws/src/atf/atf_test --package=atf_test --results-filename test_generated_atf_test_generation_config_yaml_recording_ts1_c0_r0_e0_s0_1.xml --results-base-dir "/root/target_ws/build/atf_test/test_results" /root/target_ws/build/atf_test/test_generated/atf_test_generation_config_yaml/recording_ts1_c0_r0_e0_s0_1.test 
  24: ... logging to /root/.ros/log/rostest-415ab59dcf08-10367.log
  24: [ROSUNIT] Outputting test results to /root/target_ws/build/atf_test/test_results/atf_test/rostest-test_generated_atf_test_generation_config_yaml_recording_ts1_c0_r0_e0_s0_1.xml
  24: package_name: atf_test
  24: test_generation_config_file: atf/test_generation_config.yaml
  24: test_name: ts1_c0_r0_e0_s0_1
  24: [INFO] [1626246856.725231]: waiting for smach container in test_sm to be ready...
  Error: ROR] [1626246857.820996]: Error while adding a subscriber for /reference1/x: topic parameter 'data_class' is not initialized.
  Error: ROR] [1626246857.920234]: Error while adding a subscriber for /reference1/x: topic parameter 'data_class' is not initialized.
  Error: ROR] [1626246858.021215]: Error while adding a subscriber for /reference1/x: topic parameter 'data_class' is not initialized.
  Error: ROR] [1626246858.120623]: Error while adding a subscriber for /reference1/x: topic parameter 'data_class' is not initialized.
  Error: ROR] [1626246858.220334]: Error while adding a subscriber for /reference1/x: topic parameter 'data_class' is not initialized.
  Error: ROR] [1626246858.321102]: Error while adding a subscriber for /reference1/x: topic parameter 'data_class' is not initialized.
  Error: ROR] [1626246858.420305]: Error while adding a subscriber for /reference1/x: topic parameter 'data_class' is not initialized.
  Error: ROR] [1626246858.520279]: Error while adding a subscriber for /reference1/x: topic parameter 'data_class' is not initialized.
  Error: ROR] [1626246858.620589]: Error while adding a subscriber for /reference1/x: topic parameter 'data_class' is not initialized.
  Error: ROR] [1626246858.720969]: Error while adding a subscriber for /reference1/x: topic parameter 'data_class' is not initialized.
  24: [INFO] [1626246858.745898]: ...smach container in sm_test is ready.
  24: [INFO] [1626246859.757501]: starting testblock 'testblock_small'
  24: [INFO] [1626246859.760472]: Circ
  24: [INFO] [1626246862.782934]: setting user result for testblock 'testblock_small'
  24: [INFO] [1626246862.785137]: stopping testblock 'testblock_small'
  24: [INFO] [1626246862.786943]: starting testblock 'testblock_large'
  24: [INFO] [1626246862.788568]: Circ
  24: [INFO] [1626246867.810764]: setting user result for testblock 'testblock_large'
  24: [INFO] [1626246867.812896]: stopping testblock 'testblock_large'
  24: [INFO] [1626246867.815073]: shutting down atf application
  Error: ROR] [1626246867.863748]: bad callback: <bound method ATFRecorder.global_topic_callback of <atf_core.recorder.ATFRecorder object at 0x7f32f2d746d0>>
  24: Traceback (most recent call last):
  24:   File "/opt/ros/noetic/lib/python3/dist-packages/rospy/topics.py", line 748, in _invoke_callback
  24:     cb(msg, cb_args)
  24:   File "/root/target_ws/install/atf_core/lib/python3/dist-packages/atf_core/recorder.py", line 326, in global_topic_callback
  24:     self.bag_file_writer.write_to_bagfile(name, msg, rospy.Time.now())
  24:   File "/root/target_ws/install/atf_core/lib/python3/dist-packages/atf_core/bagfile_helper.py", line 26, in write_to_bagfile
  24:     self.bag_file.write(topic, data, timestamp)
  24:   File "/opt/ros/noetic/lib/python3/dist-packages/rosbag/bag.py", line 595, in write
  24:     raise ValueError('I/O operation on closed bag')
  24: ValueError: I/O operation on closed bag
```

File: atf_core/src/atf_core/recorder.py (hunk 1 above)
Comment (fmessmer) [reply atf_core/src/atf_core/recorder.py]: done in https://github.com/floweisshardt/atf/pull/186

File: atf_core/src/atf_core/recorder.py (hunk 2 above)
Comment (fmessmer) [reply atf_core/src/atf_core/recorder.py]: done in https://github.com/floweisshardt/atf/pull/186

File: atf_core/src/atf_core/recorder.py (hunk 3 above)
Comment (fmessmer) [reply atf_core/src/atf_core/recorder.py]: done in https://github.com/floweisshardt/atf/pull/186

File: atf_core/src/atf_core/recorder.py (hunk 4 above)
Comment (fmessmer) [reply atf_core/src/atf_core/recorder.py]: done in https://github.com/floweisshardt/atf/pull/186

File: atf_core/src/atf_core/recorder.py (hunk 5 above)
Comment (fmessmer) [reply atf_core/src/atf_core/recorder.py]: done in https://github.com/floweisshardt/atf/pull/186

File: atf_core/src/atf_core/recorder.py (hunk 6 above)
Comment (fmessmer) [reply atf_core/src/atf_core/recorder.py]: done in https://github.com/floweisshardt/atf/pull/186

**Comment by fmessmer** [comment]
the extensions for `wait_timeout` from https://github.com/floweisshardt/atf/pull/186 work well
see https://github.com/mojin-robotics/kevin_atf/pull/7

however, there are some more observations in the comments above and in https://github.com/floweisshardt/atf/pull/186#pullrequestreview-706026554

File: atf_core/src/atf_core/recorder.py (hunk 3 above)
Comment (floweisshardt) [reply atf_core/src/atf_core/recorder.py]: It's not the same if you have to wait for 10 topics. I considder the `wait_timeout` to be the maximum overall time we want to wait for the test to start, not the maximum time per topic. E.g. if you have to wait 60sec for 10 topics you end up in 600sec before the test aborts.

Same for a combination of waiting for services, action, topics, tf and diagnostics. If the `wait_timeout` is set to 60s I expect the test to fail after (or at least close to) 60s, not depending on how many topics/services/tfs... are configured.

Comment (fmessmer) [reply atf_core/src/atf_core/recorder.py]: lets discuss the remaining issues and my contributions from https://github.com/floweisshardt/atf/pull/186 in person....

File: atf_core/src/atf_core/recorder.py (hunk 1 above)
Comment (floweisshardt) [reply atf_core/src/atf_core/recorder.py]: cherry-picked relevant changes from #186

File: atf_core/src/atf_core/recorder.py (hunk 2 above)
Comment (floweisshardt) [reply atf_core/src/atf_core/recorder.py]: cherry-picked relevant changes from #186 and only print latest diagnostic failures to not blow up the log output

File: atf_core/src/atf_core/recorder.py (hunk 6 above)
Comment (floweisshardt) [reply atf_core/src/atf_core/recorder.py]: cherry-picked relevant changes from #186

**Pull request merged**